### DISCUSSION TITLE
UserControl Content Property which is a Collection (WPF Equivalency)

### DISCUSSION THREAD

**Discussion by djrecipe** [AvaloniaUI/Avalonia#12580] [Q&A] [answered]
**Is it possible to bind `IList` to `UserControl.Content`?**

Description
========================================================

Hello,
I am using Avalonia's `NativeControlHost` in a UserControl and I would like to utilize the `Content` property of my UserControl to allow consumers of my UserControl to easily populate the control with custom-rendered data items. Something like this:

```XAML
<MyControl>
   <Animal>Cow</Animal>
   <Animal>Bird</Animal>
   <Animal>Horse</Animal>
</MyControl>
```
Where "Foo" is just data (collection of Animal) to help describe what the user wants me to render on the `NativeControlHost`

WPF
========================================================

In WPF I can do something similar to this:

```C#
[ContentProperty(nameof(AnimalCollection))]
public partial class GChartControl : UserControl
{
    ...
    public IList AnimalCollection
    {
        get { return (IList)GetValue(AnimalCollectionProperty); }
        set { SetValue(AnimalCollectionProperty, value); }
    }
    ...
 }
```
...and then translate each Animal into native OpenGL calls or whatever I'd like to do.

Summary
========================================================

How can I configure my UserControl to allow binding of a list of data objects to the `Content` area of the UserControl?

Cheers,
Justin Pierce

**Comment by timunie** [comment, marked as answer]
Hey @djrecipe 

See what `ITemsControl` or `ContentControl` does: 
https://github.com/AvaloniaUI/Avalonia/blob/c59e63ef6f17bdd73116df8ce177de747f894510/src/Avalonia.Controls/ItemsControl.cs#L120-L212

**Reply by djrecipe** [reply]
Thanks, I'll check it out this weekend and mark this as answer when finished :)

It's been a pleasure using your framework thus far.

**Reply by djrecipe** [reply]
Marking this as answer, but please see my suggestion below as well.

**Comment by djrecipe** [comment]
@timunie Hey, I just wanted to circle back around on this one.

Your suggestion eventually lead me to a solution, which I will outline below:

**Requirements**
I have some unique requirements. I don't actually want the "contents" to directly determine the appearance of the user control, since I am doing all of my rendering in native code. The only true thing rendered will be a single `NativeControlHost`.

I only want the "inner XAML" (aka "Contents") of my user control to be available to consumers as a sort of syntactic sugar to help describe what I should render in native code.

**Solution**
`ContentTemplate` gives me a way of parsing what the user specifies in XAML while still only rendering a singular `NativeControlHost`

_UserControl.cs_
```
this.ContentTemplate = new MyDataTemplate();
```

_MyDataTemplate.cs_
```
public Control Build(object? param)
{
   this.Items = param as MyCustomDataType; // this will be accessed and parased later on to customize and populate the native renderer :)
   var host = new MyNativeControlHost(); // always only render a singular native control host
   return host;
}
```

**Reply by timunie** [reply]
glad to hear you found a working solution ❤️ . Thx for sharing.
